compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFlattenLogicalObjectFifo.cpp

```diff
@@ -0,0 +1,101 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Iterators.h"
+
+#define DEBUG_TYPE "iree-amdaie-flatten-logicalobjectfifo"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIEFlattenLogicalObjectFifoPass
+    : public impl::AMDAIEFlattenLogicalObjectFifoBase<
+          AMDAIEFlattenLogicalObjectFifoPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect, memref::MemRefDialect>();
+  }
+
+  AMDAIEFlattenLogicalObjectFifoPass() = default;
+  AMDAIEFlattenLogicalObjectFifoPass(
+      const AMDAIEFlattenLogicalObjectFifoPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIEFlattenLogicalObjectFifoPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  ModuleOp moduleOp = getOperation();
+  IRRewriter rewriter(context);
+
+  moduleOp->walk([&](AMDAIE::LogicalObjectFifoFromMemrefOp op) {
+    MemRefType oldType = op.getMemrefType();
+    // If there is memref.cast and generates `offset: ?`, we need to use the
+    // type from memref source.
+    // TODO (vivian): check again after proper fix in
+    // DistributeCoresAndObjectFifosPass.
+    auto castOp = dyn_cast<memref::CastOp>(op.getMemref().getDefiningOp());
+    if (castOp) {
+      oldType = cast<MemRefType>(castOp.getSource().getType());
+    }
+
+    // Get linearized size and new type accordingly.
+    SmallVector<int64_t> shape = llvm::to_vector(oldType.getShape());
+    int64_t linearizedSize = 1;
+    for (auto s : shape) {
+      linearizedSize *= s;
+    }
+    MemRefType newType =
+        MemRefType::get(linearizedSize, oldType.getElementType(),
+                        MemRefLayoutAttrInterface{}, oldType.getMemorySpace());
+
+    rewriter.setInsertionPoint(op);
+    auto newLogicalObjectFifo =
+        rewriter.create<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+            rewriter.getUnknownLoc(), LogicalObjectFifoType::get(newType),
+            op.getMemref(), op.getTiles());
+    rewriter.replaceOp(op, newLogicalObjectFifo);
+
+    // Replace the access op and insert `memref.reinterpret_cast` to get to the
+    // original local shape as the objectfifo has a single type, while the DMA
+    // operations converted into objectfifos can have a different source and
+    // target type.
+    for (Operation *user : newLogicalObjectFifo->getUsers()) {
+      if (auto accessOp = dyn_cast<AMDAIE::LogicalObjectFifoAccessOp>(user)) {
+        rewriter.setInsertionPoint(accessOp);
+        auto newAccessOp = rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+            rewriter.getUnknownLoc(), newLogicalObjectFifo.getOutput(),
+            accessOp.getAccessType());
+
+        llvm::ArrayRef<int64_t> sizes = oldType.getShape();
```

Comment:
No need to store `oldType.getShape()` if there are no other uses other than the one in `rewriter.create<memref::ReinterpretCastOp>` below.